src/chainstate/stacks/db/blocks.rs

```diff
@@ -0,0 +1,2427 @@
+            let args = [&priority_chain_tip.anchored_header.block_hash().to_hex() as &dyn ToSql, &priority_chain_tip.burn_header_hash.to_hex() as &dyn ToSql];
+            let rows = query_rows::<StagingBlock, _>(&self.blocks_db, &sql, &args).map_err(Error::DBError)?;
+            if rows.len() > 0 {
+                let staging_block = rows[0].clone();
+                let microblocks = match self.find_next_staging_microblock_stream(&staging_block)? {
+                    Some(microblocks) => {
+                        return Ok(Some((staging_block, microblocks)));
+                    }
+                    None => {}
+                };
+            }
+        }
+
+        // pick a random block 
+        let sql = format!("SELECT {} FROM staging_blocks WHERE processed = 0 ORDER BY RANDOM() LIMIT 1", &row_order);
+        let rows = query_rows::<StagingBlock, _>(&self.blocks_db, &sql, NO_PARAMS).map_err(Error::DBError)?;
+        if rows.len() > 0 {
+            let staging_block = rows[0].clone();
+            let microblocks = match self.find_next_staging_microblock_stream(&staging_block)? {
+                Some(microblocks) => {
+                    return Ok(Some((staging_block, microblocks)));
+                }
+                None => {
+                    debug!("Missing parent microblock stream for {}", staging_block.anchored_block_hash.to_hex());
+                }
+            };
+        }
+
+        // no staging blocks, or no staging blocks with parent microblock streams
+        return Ok(None);
+    }
+
+    /// Process a stream of microblocks
+    /// Return the fees and burns.
+    /// TODO: if we find an invalid Stacks microblock, then punish the miner who produced it
+    fn process_microblocks_transactions<'a>(clarity_tx: &mut ClarityTx<'a>, microblocks: &Vec<StacksMicroblock>) -> Result<(u128, u128), (Error, BlockHeaderHash)> {
+        let mut fees = 0u128;
+        let mut burns = 0u128;
+        for microblock in microblocks.iter() {
+            for tx in microblock.txs.iter() {
+                let (tx_fee, tx_burns) = StacksChainState::process_transaction(clarity_tx, tx)
+                    .map_err(|e| (e, microblock.block_hash()))?;
+
+                fees = fees.checked_add(tx_fee as u128).expect("Fee overflow");
+                burns = burns.checked_add(tx_burns as u128).expect("Burns overflow");
+            }
+        }
+        Ok((fees, burns))
+    }
+
+    /// Process a single anchored block.
+    /// Return the fees and burns.
+    fn process_block_transactions<'a>(clarity_tx: &mut ClarityTx<'a>, block: &StacksBlock) -> Result<(u128, u128), Error> {
+        let mut fees = 0u128;
+        let mut burns = 0u128;
+        for tx in block.txs.iter() {
+            let (tx_fee, tx_burns) = StacksChainState::process_transaction(clarity_tx, tx)?;
+            fees = fees.checked_add(tx_fee as u128).expect("Fee overflow");
+            burns = burns.checked_add(tx_burns as u128).exp
```

Comment:
The microblocks being validated and processed here are the parent block's microblocks, right? Are they associated in the database with the child's burn_header_hash and anchored_block_hash or should they be associated with the parent's?

Reply:
Yes, they're the parent's.  They're associated in the staging microblock database with the parent's burn_header_hash and anchored_block_hash.  This isn't directly stated in the docstring for `store_staging_microblock`, but it is mentioned in `preprocess_streamed_microblock` (which validates and queues up microblocks).

I have some (uncommitted) code that shows how this works in `miner.rs` (the version in Github doesn't process microblocks yet).

Reply:
Gotcha -- it looks like at lines 1683 and 1704, it's marking the microblocks for `next_staging_block` (the child) as processed, rather than the parent, which is what it's validating.

Also, not sure why this github comment thread is appended to line 1374 (it should be line 1683, and appears so in other views -- github bug maybe?)

Reply:
> Gotcha -- it looks like at lines 1683 and 1704, it's marking the microblocks for next_staging_block (the child) as processed, rather than the parent, which is what it's validating.

Yes; that's a bug (fixed in my tree).

Reply:
Fixed in b540a484ee9e610559b3a318f2946db2ecb7d6ce